engine/evd.ml

```diff
@@ -139,6 +139,29 @@ module Abstraction = struct
   let abstract_last l = Abstract :: l
 end
 
+module Identity :
+sig
+  type t
+  val make : econstr list -> t
+  val none : unit -> t
+  val repr : named_context_val -> Filter.t -> t -> econstr list
+  val is_identity : econstr list -> t -> bool
+end =
+struct
+  type t = econstr list option ref
+  let make s = ref (Some s)
+  let none () = ref None
+  let repr sign filter s = match !s with
+  | None ->
+    let ans = Filter.filter_list filter sign.env_named_var in
+    let () = s := Some ans in
+    ans
+  | Some s -> s
```

Comment:
Can you write in a comment why the filter is invariant? To the best of my knowledge it is not, but maybe it is in the code paths of interest here. I mean, why don't use filter s, or check it conforms?

Reply:
Also, it could help a comment saying how this option type helps (which use case can take advantage of this "lazy" assignment)

Reply:
I am not sure I understand your comment. Do you mean that it could be the case that the filter doesn't match with the context? This would be a bug, since filters need to have the same length as the named context they apply to.

Reply:
It seems that repr stores in the ref something computed from the filter passed to it. Is the filter invariant? Wasn't the filter a thing to implement clear without allocating a new evar, but updating the filter?

Reply:
No, filtering implies introducing a new proxy evar. Once the evar allocated, we never touch the contents of the evarinfo.

Reply:
Ok then, but another time you will recall me why this filter is needed in the first place.